File: rust/crates/openkeyv/src/utils/compound.rs
//! Canonical compound identities for stores that do not natively support collections.

use crate::error::{Error, Result};

/// Physical key namespace used by stores sharing one database.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Subspace(String);

impl Subspace {
    pub fn new(value: impl Into<String>) -> Self {
        Self(value.into())
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn prefix(&self) -> String {
        if self.is_empty() {
            String::new()
        } else {
            format!("{}:{}", self.0.len(), self.0)
        }
    }

    pub fn scope(&self, key: &str) -> String {
        let mut scoped = self.prefix();
        scoped.push_str(key);
        scoped
    }
}

/// Return the canonical prefix for every key in `collection`.
///
/// The length is the collection's UTF-8 byte length, not its character count.
pub fn collection_prefix(collection: &str) -> String {
    format!("{}:{}", collection.len(), collection)
}

/// Encode an exact `(collection, key)` pair as one unambiguous identity.
pub fn compound_key(collection: &str, key: &str) -> String {
    let mut identity = collection_prefix(collection);
    identity.push_str(key);
    identity
}

/// Encode an exact `(collection, key)` pair inside `subspace`.
pub fn subspace_compound_key(subspace: &Subspace, collection: &str, key: &str) -> String {
    subspace.scope(&compound_key(collection, key))
}

/// Decode a canonical compound identity into its borrowed collection and key.
pub fn decompound_key(compound: &str) -> Result<(&str, &str)> {
    let (length, payload) = compound.split_once(':').ok_or_else(|| {
        Error::InvalidKey("compound identity is missing its length delimiter".to_string())
    })?;

    if length.is_empty() || !length.bytes().all(|byte| byte.is_ascii_digit()) {
        return Err(Error::InvalidKey(
            "compound identity has an invalid collection length".to_string(),
        ));
    }
    if length.len() > 1 && length.starts_with('0') {
        return Err(Error::InvalidKey(
            "compound identity has a non-canonical collection length".to_string(),
        ));
    }

    let collection_len = length.parse::<usize>().map_err(|_| {
        Error::InvalidKey("compound identity collection length is too large".to_string())
    })?;
    let collection = payload.get(..collection_len).ok_or_else(|| {
        Error::InvalidKey(
            "compound identity collection length is outside a UTF-8 boundary".to_string(),
        )
    })?;

    Ok((collection, &payload[collection_len..]))
}

/// Decode a compound identity only when it belongs to `subspace`.
pub fn subspace_decompound_key<'a>(
    subspace: &Subspace,
    compound: &'a str,
) -> Option<(&'a str, &'a str)> {
    compound
        .strip_prefix(&subspace.prefix())
        .and_then(|identity| decompound_key(identity).ok())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::error::Error;

    #[test]
    fn canonical_identity_roundtrips_exact_strings() {
        for (collection, key) in [
            ("", ""),
            ("default", "key"),
            ("a:b", "c/d"),
            ("集合", "键🔑"),
            ("*?[\\]", ":/::"),
        ] {
            let identity = compound_key(collection, key);
            assert_eq!(decompound_key(&identity).unwrap(), (collection, key));
        }
    }

    #[test]
    fn canonical_identity_uses_utf8_byte_length() {
        let identity = compound_key("集合", "key");

        assert_eq!(identity, "6:集合key");
        assert_eq!(collection_prefix("集合"), "6:集合");
    }

    #[test]
    fn canonical_identity_distinguishes_collision_pairs() {
        let left = compound_key("a:b", "c");
        let right = compound_key("a", "b:c");

        assert_ne!(left, right);
        assert_eq!(decompound_key(&left).unwrap(), ("a:b", "c"));
        assert_eq!(decompound_key(&right).unwrap(), ("a", "b:c"));
    }

    #[test]
    fn canonical_identity_preserves_case_and_unicode_normalization() {
        assert_ne!(compound_key("Users", "Key"), compound_key("users", "Key"));
        assert_ne!(compound_key("é", "key"), compound_key("e\u{301}", "key"));
    }

    #[test]
    fn collection_prefix_matches_only_the_exact_collection_frame() {
        let prefix = collection_prefix("a:b");

        assert!(compound_key("a:b", "key").starts_with(&prefix));
        assert!(!compound_key("a", "b:key").starts_with(&prefix));
    }

    #[test]
    fn malformed_identity_is_rejected() {
        for identity in [
            "",
            ":key",
            "x:key",
            "01:akey",
            "2:a",
            "1:ékey",
            "999999999999999999999999999999999999999999999999999999999999999999999999999999:key",
        ] {
            assert!(matches!(
                decompound_key(identity),
                Err(Error::InvalidKey(_))
            ));
        }
    }

    #[test]
    fn subspace_roundtrips_and_keeps_empty_layout_compatible() {
        let empty = Subspace::default();
        assert_eq!(subspace_compound_key(&empty, "users", "key"), "5:userskey");
        assert_eq!(
            subspace_decompound_key(&empty, "5:userskey"),
            Some(("users", "key"))
        );

        let scoped = Subspace::new("租户A");
        let identity = subspace_compound_key(&scoped, "users", "key");
        assert_eq!(identity, "7:租户A5:userskey");
        assert_eq!(
            subspace_decompound_key(&scoped, &identity),
            Some(("users", "key"))
        );
        assert_eq!(
            subspace_decompound_key(&Subspace::new("other"), &identity),
            None
        );
        assert_eq!(subspace_decompound_key(&scoped, "7:租户Abad"), None);
    }
}

File: python/src/openkeyv/_utils/compound.py
"""Utilities for canonical compound identities and explicit prefixes."""

from collections.abc import Sequence

from openkeyv._utils.beartype import bear_enforce

DEFAULT_PREFIX_SEPARATOR = "__"


@bear_enforce
def collection_prefix(collection: str) -> str:
    """Return the canonical length-prefixed identity for a collection."""
    return f"{len(collection.encode('utf-8'))}:{collection}"


@bear_enforce
def compound_key(collection: str, key: str) -> str:
    """Encode an exact ``(collection, key)`` pair as one canonical identity."""
    return f"{collection_prefix(collection)}{key}"


@bear_enforce
def uncompound_key(compound: str) -> tuple[str, str]:
    """Decode a canonical compound identity into its collection and key."""
    length_text, delimiter, payload = compound.partition(":")
    if delimiter == "":
        msg = f"String {compound} is not a canonical compound identifier"
        raise TypeError(msg) from None
    if not length_text or not all("0" <= character <= "9" for character in length_text):
        msg = f"String {compound} has an invalid collection length"
        raise TypeError(msg) from None
    if len(length_text) > 1 and length_text.startswith("0"):
        msg = f"String {compound} has a non-canonical collection length"
        raise TypeError(msg) from None

    try:
        collection_length = int(length_text)
    except ValueError as error:
        msg = f"String {compound} has an invalid collection length"
        raise TypeError(msg) from error

    payload_bytes = payload.encode("utf-8")
    if len(payload_bytes) < collection_length:
        msg = f"String {compound} has a collection length outside the payload"
        raise TypeError(msg) from None

    try:
        collection = payload_bytes[:collection_length].decode("utf-8")
        key = payload_bytes[collection_length:].decode("utf-8")
    except UnicodeDecodeError as error:
        msg = f"String {compound} has a collection length outside a UTF-8 boundary"
        raise TypeError(msg) from error

    return collection, key


def prefix_key(key: str, prefix: str, separator: str | None = None) -> str:
    """Add an explicit separator-based prefix to a key."""
    separator = DEFAULT_PREFIX_SEPARATOR if separator is None else separator
    if separator == "":
        msg = "Separator must not be empty"
        raise ValueError(msg)
    return f"{prefix}{separator}{key}"


def unprefix_key(key: str, prefix: str, separator: str | None = None) -> str:
    """Remove an explicit separator-based prefix from a key."""
    separator = DEFAULT_PREFIX_SEPARATOR if separator is None else separator
    if separator == "":
        msg = "Separator must not be empty"
        raise ValueError(msg)
    if not key.startswith(prefix + separator):
        msg = f"Key {key} is not prefixed with {prefix}{separator}"
        raise ValueError(msg)
    return key[len(prefix + separator) :]


def prefix_collection(collection: str, prefix: str, separator: str | None = None) -> str:
    """Add an explicit separator-based prefix to a collection name."""
    separator = DEFAULT_PREFIX_SEPARATOR if separator is None else separator
    if separator == "":
        msg = "Separator must not be empty"
        raise ValueError(msg)
    return f"{prefix}{separator}{collection}"


def unprefix_collection(collection: str, prefix: str, separator: str | None = None) -> str:
    """Remove an explicit separator-based prefix from a collection name."""
    separator = DEFAULT_PREFIX_SEPARATOR if separator is None else separator
    if separator == "":
        msg = "Separator must not be empty"
        raise ValueError(msg)
    if not collection.startswith(prefix + separator):
        msg = f"Collection {collection} is not prefixed with {prefix}{separator}"
        raise ValueError(msg)
    return collection[len(prefix + separator) :]


def get_collections_from_compound_keys(compound_keys: Sequence[str]) -> list[str]:
    """Return the unique collections represented by canonical compound keys."""
    return list({collection for collection, _ in (uncompound_key(compound) for compound in compound_keys)})


def get_keys_from_compound_keys(compound_keys: Sequence[str], collection: str) -> list[str]:
    """Return canonical compound keys belonging to ``collection``."""
    return [key for key_collection, key in (uncompound_key(compound) for compound in compound_keys) if key_collection == collection]
